File: RB101_RB109_small_problems/easy_8/all_substrings.rb
=begin

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: string
-  Output: array of all substrings of input string

---

**Problem Domain:**
Nothing specific.
---

**Implicit Requirements:**
- Array should be ordered by where in the string the substring begins.
- Should use the #leading_substrings method from previous exercise.
---

**Clarifying Questions:**

1. None, example and problem statement make requirements clear.

---

**Mental Model:**

Iterate from 0 to the input string length - 1 (i = iteration number)
  Create leading substrings starting at string index i. Add to accumulation array.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

substrings('abcde') == [
  'a', 'ab', 'abc', 'abcd', 'abcde',
  'b', 'bc', 'bcd', 'bcde',
  'c', 'cd', 'cde',
  'd', 'de',
  'e'
]

---

Data Structure
--------------

- Use arrays to store substrings.

---

Algorithm
---------
- Initialize empty all_substrings array.
- Iterate from 0 to the input string length - 1 (i = iteration number)
  - Create leading substrings starting at string index i. Add to all_substrings array.
- Return all_substrings array.

Code
----

=end

def leading_substrings(string)
  result = []
  0.upto(string.size - 1) do |index|
    result << string[0..index]
  end
  result
end

def substrings(string)
  all_substrings = []
  string.length.times do |index|
    all_substrings += leading_substrings(string[index..-1])
  end
  all_substrings
end

substrings('abcde') == [
  'a', 'ab', 'abc', 'abcd', 'abcde',
  'b', 'bc', 'bcd', 'bcde',
  'c', 'cd', 'cde',
  'd', 'de',
  'e'
]

File: JS210_small_problems/list_processing/all_substrings.js
function leadingSubstrings(str) {
  let substrings = [];

  for (let length = 1; length <= str.length; length += 1) {
    substrings.push(str.slice(0, length));
  }

  return substrings;
}

// function substrings(str) {
//   let substringsArray = [];
//   for (let index = 0; index < str.length; index += 1) {
//     let partialStr = str.slice(index);
//     substringsArray = substringsArray.concat(leadingSubstrings(partialStr));
//   }

//   return substringsArray;
// }

console.log(substrings('abcde'));

// // returns
// [ "a", "ab", "abc", "abcd", "abcde",
//   "b", "bc", "bcd", "bcde",
//   "c", "cd", "cde",
//   "d", "de",
//   "e" ]

// Further Exploration

function substrings(str) {
  let partialStrs = str.split('').map((char, index) => str.slice(index));

  return partialStrs.map(leadingSubstrings).flat();
}